### ISSUE TITLE
Find plan for component extensibility without inheritance

### ISSUE THREAD

**Issue by lukebayes** [waybeams/waybeams#240]
We cannot extend a base component and mutate functionality in Go.

In an attempt to implement TextInput.Text() in such a way that it will use the optional Placeholder() value if Text() is an empty string, I discovered how terribly I misunderstood Go's composition vs inheritance.

An embedded struct can appropriately operate over scalar values as directed by an embedding struct via method calls. If an embedded struct calls any method on itself or passes itself as an argument to any function, the outer embedding struct is no longer available. Operations like, SetWidth(w) or SetID() work fine. The failure mode occurs whenever an abstract implementation method needs to operate on the bound instance by reference. For example, calling any other method on the current instance.

Looking for a beautiful solution that allows people to enhance base component functionality with concrete features while sticking to the spirit of the Open/Closed principle.

Should concrete control authors focus on configuring instances of the existing Control or Defining new Controls as types?

Some smells and requirements in no particular order:

* Should guide authors away from solutions like the currently deep embed hierarchy on NanoWindow -> GlfwWindow -> Control
* Must have a clear, easy to understand pattern for developing concrete functionality
* Must avoid confusion for new authors/gophers related to Embed vs Inherit
* Should feel like Idiomatic Go, not something new

**Comment by lukebayes** [comment]
We cannot extend a base component and mutate functionality in Go.

In an attempt to implement TextInput.Text() in such a way that it will use the optional Placeholder() value if Text() is an empty string, I discovered how terribly I misunderstood Go's composition vs inheritance.

Need to find a beautiful solution that allows people to enhance base component functionality with concrete features while sticking to the spirit of the Open/Closed principle.

**Comment by lukebayes** [comment]
# Failed Attempt: [Simple objects, module function interface](https://github.com/waybeams/waybeams/tree/lbj-attempt-01)

In file: `ui/compose/composable.go`

```go

type Composite struct {
	children []*Composite
	parent []*Composite
}

func AddChild(parent, child *Composite) int {
	if parent == child {
		panic("Cannot add Composite to self")
	}
	parent.children = append(parent.children, child)
	child.parent = parent
	return -1
}
```
...more of the Composable interface implemented in this fashion.

Then in file, `ui/control/control.go`
```go
type Control struct {
	compose *compose.Composite
}

func New() *Control {
	return &Control{}
}

func AddChild(parent *Control, child *Control) int {
	// Delegation to the compose surface fails because return values are 
	// not usable.
	return compose.AddChild(parent.compose, child.compose)
}

func ChildAt(parent *Control, index int) *Control {
	return compose.ChildAt(parent.compose, index)
}
```
## Pros:
* Struct will be a very simple, serializable bag of state
* Functionality can be provided/changed in a variety of modules
* Any concrete type can compose any subset of the displayable models

## Cons:
* Interface enumeration: More difficult to figure out what methods are available on a given struct.
* Multi-directional references: When the Control composes the Composable model, we need to have a back reference from Composable to Control in order to go from a list of composable model children to a list of Control children. This fails in both go import statements and feels wrong.
* Functions use concrete type, so desired polymorphism is not available, though concrete types can compose a collection of these primitives. If that's how it works, it seems like the concrete types should just compose the struct along with the struct's existing interface.

**Comment by lukebayes** [comment]
# Attempt 2: Lifecycle methods and extensible state storage

* Stop exporting Component
* Interfaces continue to operate on Displayables, some might get more specific
* Introduce lifecycle callback handlers (e.g., OnCreated, OnConfigured, OnLayoutCompleted, OnDrawn)
* Update Component.Data to a map of string keys and struct values that will allow concrete component configurations to define and manipulate values on Data.

Cons:
* Concrete components cannot introduce new methods on the Component struct or entities that appear to be instances of whatever they represent
* Leaves component authors without composition OR inheritance, just a morass of configuration.

**Comment by lukebayes** [comment]
# Conclusion

This means that:

## RULE 1: We cannot embed structs that call methods from within their defined methods or that pass themselves as references to external methods.

If only a subset of methods are moved off of the interface, it's hard for me to come up with a reasonable heuristic that would allow component authors or consumers to accurately guess where to find a given method. Is it on the instance or on the module? Why here and not there?

## RULE 2: If RULE 1 stands, functions that operate over structs _that will be embedded_ MUST be defined on a module, not the struct.

## RULE 3: If RULE 2 stands, and because function arguments must be typed and structs may be embedded, function receivers should expect interfaces, not concrete structs.

## RULE 4: If RULE 3 stands, all fields should be accessed via accessors, which violates *RULE 2*.

## RULE 5: If RULES 1-4 stand, NEVER, EVER embed a struct using Go syntactic bullshit.

**Comment by lukebayes** [comment]
Fixed in #249 

```golang
Foo(Width(23), Height(25),
  Child(Bar()),
  Child(Baz()))
```
